wcfsetup/install/files/lib/data/blacklist/entry/BlacklistEntry.class.php

```diff
@@ -0,0 +1,132 @@
+<?php
+namespace wcf\data\blacklist\entry;
+use wcf\data\DatabaseObject;
+use wcf\system\database\util\PreparedStatementConditionBuilder;
+use wcf\system\WCF;
+use wcf\util\UserUtil;
+
+/**
+ * Represents a blacklist entry.
+ * 
+ * @author      Alexander Ebert
+ * @copyright   2001-2019 WoltLab GmbH
+ * @license     GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @package     WoltLabSuite\Core\Data\Blacklist\Entry
+ * 
+ * @property-read string $type One of 'email', 'ipv4', 'ipv6' or 'username'
+ * @property-read string $hash SHA256 hash of the original value
+ * @property-read int $lastSeen Timestamp of the last report, derivation is up to 24hrs
+ * @property-read int $occurrences Number of times this value was reported, capped at 32,767
+ * @since 5.2
+ */
+class BlacklistEntry extends DatabaseObject {
+	/**
+	 * @param string $username
+	 * @param string $email
+	 * @param string $ipAddress
+	 * @return string[]
+	 */
+	public static function getMatches($username, $email, $ipAddress) {
+		if (BLACKLIST_SFS_USERNAME === 'skip' && BLACKLIST_SFS_EMAIL_ADDRESS === 'skip' && BLACKLIST_SFS_IP_ADDRESS === 'skip') {
+			return false;
+		}
+		
+		$conditions = new PreparedStatementConditionBuilder(true, 'OR');
+		if (BLACKLIST_SFS_USERNAME) {
+			$conditions->add('(type = ? AND hash = ?)', ['username', self::getHash($username)]);
+		}
+		if (BLACKLIST_SFS_EMAIL_ADDRESS) {
+			$conditions->add('(type = ? AND hash = ?)', ['email', self::getHash($email)]);
+		}
+		if (BLACKLIST_SFS_IP_ADDRESS) {
+			UserUtil::convertIPv6To4($ipAddress);
+			if ($ipAddress) {
+				if (filter_var($ipAddress, FILTER_VALIDATE_IP, FILTER_FLAG_IPV4) !== false) {
+					$conditions->add('(type = ? AND hash = ?)', ['ipv4', self::getHash($ipAddress)]);
+				}
+				else {
+					$parts = explode(':', $ipAddress);
+					
+					// StopForumSpam uses the first two to four segments of an IPv6 address.
+					$ipv6TwoParts = self::getHash("{$parts[0]}:{$parts[1]}::");
+					$ipv6ThreeParts = self::getHash("{$parts[0]}:{$parts[1]}:{$parts[2]}::");
+					$ipv6FourParts = self::getHash("{$parts[0]}:{$parts[1]}:{$parts[2]}:{$parts[3]}::");
+					
+					$conditions->add('(type = ? AND hash IN (?))', ['ipv6', [$ipv6TwoParts, $ipv6ThreeParts, $ipv6FourParts]]);
+				}
+			}
+		}
+		
+		$sql = "SELECT  type, occurrences
+			FROM    wcf".WCF_N."_blacklist_entry
+			".$conditions;
+		$statement = WCF::getDB()->prepareStatement($sql);
+		$statement->execute($conditions->getParameters());
+		$matches = [];
+		while ($row = $statement->fetchArray()) {
+			if (self::isMatch($row['type'], $row['occurrences'])) {
+				$matches[] = ($row['type'] === 'ipv4' || $row['type'] === 'ipv6') ? 'ip' : $row['type'];
+			}
+		}
+		
+		return $matches;
+	}
+	
+	protected static function getHash($string) {
+		return hex2bin(hash('sha256', $string));
```

Comment:
```suggestion
		return hash('sha256', $string, true);
```
The `$raw` parameter is equivalent to `hex2bin`.